### PR TITLE
ci: stabilize workflows

### PR DESCRIPTION
**Pull request by lenaschoenburg**
This PR combines multiple improvements and has two goals:
1. Stabilize CI
2. Improve maintainability

To achieve the first goal of stabilizing CI:
1. Use 16-core self-hosted runners. These seem to be more reliable and are less contended.
2. Disable Testcontainer Cloud. This removes one potential source of flakiness.
3. Run much less jobs overall. Every job had the potential to break due to external dependencies so having less jobs decreases the chance to encounter those.
4. Run unit tests on self hosted runners where we can use our own caching maven repository.
5. Fail faster. Timeouts are decreased, previous workflow runs are cancelled and integration tests fail as soon as one test fails. This frees up resources and ensures that one failing PR doesn't block other PRs for too long.

To improve maintainability:
1. Merge code quality and test workflows. These used the same triggers anyway and it's nice to have all required checks in one file.
2. Run all unit tests in one job. This makes looking for failing checks in the GH UI easier
3. Merge multiple integration test jobs. Figuring out which tests ran where was tricky so now we just have one job that runs integration tests.

This has a couple of drawbacks:
1. Time to failure increases: Previously some test failures were found in 1-2 minutes, now we need to wait for all unit tests (>15 minutes) before checks fail.
2. Workflow run time increased by about 3 minutes, primarily coming from the unit test job.
3. Less granular retry.
4. Increased chances of failure due to using 16-core runners for integration tests and not using Testcontainer Cloud.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -71,78 +70,27 @@ jobs:
...
           go: false
-          maven-extra-args: -am -pl ${{ matrix.project }}
       - name: Create build output log file
         run: echo "BUILD_OUTPUT_FILE_PATH=$(mktemp)" >> $GITHUB_ENV
       - name: Maven Test Build
         run: >
           mvn -B --no-snapshot-updates
           -D skipITs -D skipChecks
-          -pl ${{ matrix.project }}
+          -D junitThreadCount=16
```
Comment (megglos) [inline .github/workflows/test.yml:92]: assuming there are modules that can run in parallel would adding `-T 1C` be worth it for speed-up?

Comment (oleschoenburg) [reply .github/workflows/test.yml:92]: Maybe, yeah :+1: I've tried to max out CPU via JUnit first but `RandomizedRaftTest` is taking 5 minutes and blocking progress for all others.

Comment (oleschoenburg) [reply .github/workflows/test.yml:97]: Done, it removed around 3 minutes

**Comment by npepinpe** [comment]
Can we have a short chat in the TR about solutions? Generally some of these are good ideas, but I have some ideas/suggestions (e.g. split unit tests across a fixed number of groups), and questions (e.g. why is Testcontainers Cloud a source of flakiness?).

**Comment by Zelldon** [comment]
Some potential alternative: way for now would be also to run bors with jenkins https://github.com/camunda/zeebe/commit/657be1ff9fbf4d3d2aa9902bf72f826884886889

**Review by npepinpe — APPROVED** [review]
yolo

**Review by megglos — APPROVED** [review]
looks good to me let's see it in action 🤞

**Comment by deepthidevaki** [comment]
bors r-

already merged

**Pull request merged**